scripts/utils.py

```diff
@@ -0,0 +1,322 @@
+    logging.info(pca.components_)
+
+    principalComponents = pca.fit_transform(x)
+
+    return principalComponents
+
+
+def _plot_pca_data(principalComponents, Ly, Uy, ax):
+    '''
+    Helper function for plot_pca that plots data for a given axis.
+    Inputs:
+    principalComponents: ndarray of shape (n_samples, n_components).
+    Ly: class labels for labeled data.
+    Uy: labels for unlabeled data (all labels should be -1).
+    ax: matplotlib-axis to plot on.
+    '''
+
+    # only saving colors for binary classification with unlabeled instances
+    col_dict = {-1: 'tab:gray', 0: 'tab:orange', 1: 'tab:blue'}
+
+    for idx, color in col_dict.items():
+        indices = np.where(np.append(Ly, Uy, axis=0) == idx)[0]
+        ax.scatter(principalComponents[indices, 0],
+                   principalComponents[indices, 1],
+                   c=color,
+                   label='class '+str(idx))
+    return ax
+
+
+def plot_pca(principalComponents, Ly, Uy, filename, n=2):
+    '''
+    A function for computing and plotting n-dimensional PCA.
+    Inputs:
+    principalComponents: ndarray of shape (n_samples, n_components).
+    Ly: class labels for labeled data.
+    Uy: labels for unlabeled data (all labels should be -1).
+    filename: filename for saved plot.
+        The file must be saved with extension .joblib.
+        Added to filename if not included as input.
+    n: number of singular values to include in PCA analysis.
+    '''
+
+    plt.rcParams.update({'font.size': 20})
+
+    alph = ["A", "B", "C", "D", "E", "F", "G", "H",
+            "I", "J", "K", "L", "M", "N", "O", "P",
+            "Q", "R", "S", "T", "U", "V", "W", "X",
+            "Y", "Z"]
+    jobs = alph[:n]
+
+    # only one plot is needed for n=2
+    if n == 2:
+        fig, ax = plt.subplots(figsize=(10, 8))
+        ax.set_xlabel('PC '+jobs[0], fontsize=15)
+        ax.set_ylabel('PC '+jobs[1], fontsize=15)
+        ax = _plot_pca_data(principalComponents, Ly, Uy, ax)
+        ax.grid()
+        ax.legend()
+    else:
+        fig, axes = plt.subplots(n, n, figsize=(15, 15))
+        for row in range(axes.shape[0]):
+            for col in range(axes.shape[1]):
+                ax = axes[row, col]
+                # blank label plot
+                if row == col:
+                    ax.tick_params(
+                        axis='both', which='both',
+                        bottom='off', top='off',
+                        labelbottom='off',
+                        left='off', right='off',
+                        labelleft='off'
+                    )
+                    ax.text(0.5, 0.5, jobs[row], horizontalalignment='center')
+                # PCA results
+                else:
+                    ax = _plot_pca_data(principalComponents, Ly, Uy, ax)
+
+    if filename[-4:] != '.png':
+        filename += '.png'
+    fig.tight_layout()
+    fig.savefig(filename)
+
+
+def plot_cf(testy, predy, title, filename):
```

Comment:
@gonuke, per our conversations from S/W, do you still concur that we can ignore testing these PCA/plotting functions?

Reply:
Yes - but is there a way to exclude them from the denominator of the coverage testing?

Reply:
Turns out, Coveralls supports adding `# pragma: no cover` to blocks of code, and it will ignore it. Adding to the plotting functions above.